cachi2/core/package_managers/pip.py

```diff
@@ -1370,20 +1453,140 @@ def _split_hashes_from_options(cls, options: list[str]) -> tuple[list[str], list
+    download_info["requirement_file"] = str(requirements_file.file_path.subpath_from_root)
+    download_info["hash_verified"] = False
+
+    def _hash_verify(path: Path, checksum_info: Iterable[ChecksumInfo]) -> bool:
+        verified: bool = False
+        try:
+            must_match_any_checksum(path, checksum_info)
+            verified = True
+        except PackageRejected:
+            path.unlink()
+            log.warning("Download '%s' was removed from the output directory", path.name)
+        return verified
+
+    if dpi:
+        if dpi.should_verify_checksums():
+            download_info["hash_verified"] = _hash_verify(dpi.path, dpi.checksums_to_verify)
+
+        if dpi.package_type == "sdist":
+            _check_metadata_in_sdist(dpi.path)
+
+    else:
+        if require_hashes or req.kind == "url":
+            hashes = req.hashes or [req.qualifiers.get("cachito_hash", "")]
+            download_info["hash_verified"] = _hash_verify(
+                download_info["path"], list(map(_to_checksum_info, hashes))
+            )
+
+    log.info(
+        "-- Successfully processed '%s' in path '%s'",
+        req.download_line,
+        download_info["path"].relative_to(pip_deps_dir.root),
+    )
+
+    return download_info
+
+
+def _process_pypi_req(
+    req: PipRequirement,
+    requirements_file: PipRequirementsFile,
+    require_hashes: bool,
+    pip_deps_dir: RootedPath,
+    allow_binary: bool,
+) -> list[dict[str, Any]]:
+    download_infos: list[dict[str, Any]] = []
+
+    def _process_req_wrapper(
+        DPIs: list[DistributionPackageInfo], log_string: str
+    ) -> list[dict[str, Any]]:
+        infos: list[dict[str, Any]] = []
+        for dpi in DPIs:
+            log.info(log_string, dpi.path.name)
+            infos.append(
+                _process_req(
+                    req, requirements_file, require_hashes, pip_deps_dir, dpi.download_info, dpi=dpi
+                )
+            )
+        return infos
+
+    artifacts: list[DistributionPackageInfo] = _process_package_distributions(
+        req, pip_deps_dir, allow_binary
+    )
+    artifacts_to_check: list[DistributionPackageInfo] = [a for a in artifacts if a.path.exists()]
+    artifacts_to_download: list[DistributionPackageInfo] = [
+        a for a in artifacts if not a.path.exists()
+    ]
+
+    files: dict[str, Union[str, PathLike[str]]] = {
+        dpi.url: dpi.path for dpi in artifacts_to_download
+    }
+    asyncio.run(async_download_files(files, get_config().concurrency_limit))
+
+    download_infos.extend(
+        _process_req_wrapper(artifacts_to_check, "-- Checking existing pypi artifact '%s'")
+    )
+    download_infos.extend(
+        _process_req_wrapper(artifacts_to_download, "-- Processing pypi artifact to download '%s'")
+    )
+
+    return download_infos
+
+
+def _process_vcs_req(
+    req: PipRequirement, pip_deps_dir: RootedPath, **kwargs: Any
+) -> dict[str, Any]:
+    log.info("-- Processing vcs repo '%s'", req.url)
```

Comment:
Do these symbols have any special meaning? ("--" or "*")
I find that kind of ugly when all other messages start immediately after the log level.

```
2024-03-18 20:58:13,087 INFO Extracting metadata from pyproject.toml
2024-03-18 20:58:13,088 INFO Resolved package name: 'wheels-of-fortune'
2024-03-18 20:58:13,089 INFO Resolved package version: '0.1.0'
2024-03-18 20:58:13,090 INFO At least one dependency uses the --hash option, will require hashes
2024-03-18 20:58:13,090 INFO * Processing requirement line 'pyicu==2.12'
2024-03-18 20:58:13,713 INFO -- Processing pypi artifact to download 'PyICU-2.12.tar.gz'
2024-03-18 20:58:13,713 INFO Verifying checksums of PyICU-2.12.tar.gz
2024-03-18 20:58:13,718 INFO -- Successfully processed 'pyicu==2.12' in path 'deps/pip/PyICU-2.12.tar.gz'
2024-03-18 20:58:13,718 INFO * Finished processing requirement line 'pyicu==2.12
```

Reply:
I don't know **exactly** what you mean by "special", but, *textually*, the '*' marks the start/end of processing a single requirement, and '--' marks the start/end of processing a single artifact of an enclosing requirement.

This is all in order to demarcate different actions, to make it **easier** to read the logs. It is *extremely* difficult to figure out what's happening when one is looking at an undifferentiated wall of text.

Reply:
TL;DR - [This](https://privatebin.corp.redhat.com/?6d4ecc07257cc34d#9Gzthnx96JsbR5xV7dSVM1jLT8Tgbw6g83jG18VVfrxB) is the cachi2 log, with `allow_binary` enabled, from a *real-world* project ([StableCascade](https://github.com/Stability-AI/StableCascade)) which has many complex dependencies.

I ran it once and interrupted it, then ran it again and captured the log.

If you search for 'click', you can clearly see that above that point in that log, cachi2 is *checking*; below that, it's *downloading*

Yes, it's a lot of log, but the line breaks and "line headers" make it easy to see what's going on.

Reply:
Oh, I didn't understand that that's what you meant - I'll look at changing some to DEBUG.